packages/forms/components/TextInputWithIcon.tsx

```diff
@@ -0,0 +1,132 @@
+import { cx } from "emotion";
+import * as React from "react";
+
+import { TextInput, TextInputAppearance, TextInputProps } from "./TextInput";
+
+import { inputAppearances, inputContainer, iconEnd } from "../style";
+import { flex, flexItem, padding } from "../../shared/styles/styleUtils";
+
+export interface TextInputWithIconProps extends TextInputProps {
+  /**
+   * icon to display at the start of this TextInput.
+   */
+  iconStart?: React.ReactNode;
+  /**
+   * icon to display at the end of this TextInput.
+   */
+  iconEnd?: React.ReactNode;
+}
+
+export interface TextInputWithIconState {
+  hasFocus: boolean;
+}
+
+export class TextInputWithIcon extends TextInput<
+  TextInputWithIconProps,
+  TextInputWithIconState
+> {
+  public static defaultProps: Partial<TextInputWithIconProps> = {
+    type: "text",
+    appearance: TextInputAppearance.Standard,
+    showInputLabel: true
+  };
+  constructor(props) {
+    super(props);
+
+    this.inputOnFocus = this.inputOnFocus.bind(this);
+    this.inputOnBlur = this.inputOnBlur.bind(this);
+
+    this.state = {
+      hasFocus: false
+    };
+  }
+
+  protected getInputAppearance(): string {
+    if (this.props.disabled) {
+      return "disabled";
+    }
+    if (this.state.hasFocus) {
+      return `${this.props.appearance}-focus`;
+    }
+    return this.props.appearance;
+  }
+
+  protected getInputElementProps() {
+    let inputProps = super.getInputElementProps();
+    inputProps.onFocus = this.inputOnFocus;
+    inputProps.onBlur = this.inputOnBlur;
+    return inputProps;
+  }
+
+  protected getIconStartContent() {
+    if (!this.props.iconStart) {
+      return null;
+    }
+    return (
+      <span
+        className={cx(
+          padding("right", "xs"),
+          flex({ align: "center", justify: "center" }),
+          flexItem("shrink")
+        )}
+      >
+        {this.props.iconStart}
+      </span>
+    );
+  }
+
+  protected getIconEndContent() {
+    if (!this.props.iconEnd) {
+      return null;
+    }
+    return (
+      <span
+        className={cx(
+          padding("left", "xs"),
+          flex({ align: "center", justify: "center" }),
+          flexItem("shrink"),
+          iconEnd
+        )}
```

Comment:
Add `flush("left")`.
I thought I was being "helpful" by adding gutters for flex items, but it seems to be hurting more than helping in a lot of cases...